### PR TITLE
topology: add dmic pipeline to apl nocodec topology

### PR DESCRIPTION
**Pull request by ranj063**
Add the two DMIC capture pipelines to the apl nocodec topology.

### CODE REVIEW

**Comment by ranj063** [comment]
@plbossart @xiulipan @keyonjie Could you please help review?

**Comment by keyonjie** [comment]
@ranj063 why are you adding DMIC to nocodec tplg? Do you have any target board to use for it?

**Comment by ranj063** [comment]
@plbossart. Yes let me fix the samples to 16. Should the sample rate also change?

I am not sure how to configure the dai for dual usages? @keyonjie any ideas?

as for the NoCodec-6 and NoCodec-7 names, they're coming from the way the nocodec driver names the dai links. The noccodec driver sets up the BE link names as Nocodec-i, with i being the index of the link in the BE dai array.

**Comment by xiulipan** [comment]
@ranj063 @lgirdwood @plbossart 
Test with this patch and add the HP buuffer size.
Find few question:
1. these two dai_get function used different variables  https://github.com/thesofproject/sof/blob/d696d8f8bad1490575a8293e0982786300eb52db/src/audio/dai.c#L194 and https://github.com/thesofproject/sof/blob/d696d8f8bad1490575a8293e0982786300eb52db/src/ipc/handler.c#L492
It is easy for SSP to keep this align. But for DMIC, the index in the array starts from 0 again.
Either we could not find the comp dai or we could not set the dai config.
This may need some kernel side and firmware side alignment.

2. In tplg, DMIC DAI_INDEX starts from 0, please see my inline comments.

I found 1 error after fix 2.

**Comment by xiulipan** [comment]
memory size increase https://github.com/thesofproject/sof/pull/132

**Comment by xiulipan** [comment]
kernel topology parse change PR:
https://github.com/thesofproject/linux/pull/42

**Comment by plbossart** [comment]
@xiulipan I couldn't figure what you are trying to explain. It looks like there is a confusion between two id fields (unfortunately classic with topology since everything is called id) but I don't get your point really, nor how you fix things. can you elaborate on your findings?

**Comment by xiulipan** [comment]
@plbossart 
let me paste all codes here without link.

function used sof_ipc_dai_config structure(in src/ipc/handler.c)
```

static int ipc_dai_config(uint32_t header)
{
struct sof_ipc_dai_config *config = _ipc->comp_data;
....
dai = dai_get(config->type, config->id);
....
}

```

function used sof_ipc_comp_dai  structure(in src/audio/dai.c)

```
static struct comp_dev *dai_new(struct sof_ipc_comp *comp)
{
	struct comp_dev *dev;
	struct sof_ipc_comp_dai *dai;
	struct sof_ipc_comp_dai *ipc_dai = (struct sof_ipc_comp_dai *)comp;
.....
	dd->dai = dai_get(dai->type, dai->index);
	if (dd->dai == NULL) {
		trace_dai_error("eDg");
		goto error;
	}
....
	return NULL;
}

```
This two functions (ipc_dai_config and dai_new) should get the same dai from the dai_get function( in file 
src/platform/apollolake/dai.c)
```
struct dai *dai_get(uint32_t type, uint32_t index)
{
	int i;

	if (type == SOF_DAI_INTEL_SSP) {
		for (i = 0; i < ARRAY_SIZE(ssp); i++) {
			if (ssp[i].type == type && ssp[i].index == index)
				return &ssp[i];
		}
	}
	if (type == SOF_DAI_INTEL_DMIC) {
		for (i = 0; i < ARRAY_SIZE(dmic); i++) {
			if (dmic[i].type == type && dmic[i].index == index)
				return &dmic[i];
		}
	}
	return NULL;
}
```
The structure dmic is here:

```
static struct dai dmic[2] = {
	/* Testing idea if DMIC FIFOs A and B to access the same microphones
	 * with two different sample rate and PCM format could be presented
	 * similarly as SSP0..N. The difference however is that the DMIC
	 * programming is global and not per FIFO.
	 */

	/* Primary FIFO A */
	{
		.type = SOF_DAI_INTEL_DMIC,
		.index = 0,
		.plat_data = {
			.base = DMIC_BASE,
			.irq = IRQ_EXT_DMIC_LVL5(0),
			.fifo[SOF_IPC_STREAM_PLAYBACK] = {
				.offset = 0, /* No playback */
				.handshake = 0,
			},
			.fifo[SOF_IPC_STREAM_CAPTURE] = {
				.offset = DMIC_BASE + OUTDATA0,
				.handshake = DMA_HANDSHAKE_DMIC_CH0,
			}
		},
		.ops = &dmic_ops,
	},
	/* Secondary FIFO B */
	{
		.type = SOF_DAI_INTEL_DMIC,
		.index = 1,
		.plat_data = {
			.base = DMIC_BASE,
			.irq = IRQ_EXT_DMIC_LVL5(0),
			.fifo[SOF_IPC_STREAM_PLAYBACK] = {
				.offset = 0, /* No playback */
				.handshake = 0,
			},
			.fifo[SOF_IPC_STREAM_CAPTURE] = {
				.offset = DMIC_BASE + OUTDATA1,
				.handshake = DMA_HANDSHAKE_DMIC_CH1,
			}
		},
		.ops = &dmic_ops,
	}
};
```

So from the above codes, we can tell that the dai used type and **index** NOT ID to differentiate from each other. So we should use index in ipc_dai_config but not ID, the filed name in sof_ipc_dai_config maybe some error at first. And this TYPE and INDEX are all vendor date we defined in our Firmware array.

In this PR
https://github.com/thesofproject/linux/pull/42
I just make some unification for these two user of the dai.

Maybe I should change id => index since we have some name misusing here.
```
/* general purpose DAI configuration */
struct sof_ipc_dai_config {
	struct sof_ipc_hdr hdr;
	enum sof_ipc_dai_type type;
	uint32_t id;	/* physical number if more than 1 of this type */ 
// change this id to index to make alignment with other usage.

	/* physical protocol and clocking */
	uint16_t format;	/* SOF_DAI_FMT_ */
	uint16_t reserved;	/* alignment */

	/* HW specific data */
	union {
		struct sof_ipc_dai_ssp_params ssp;
		struct sof_ipc_dai_hda_params hda;
		struct sof_ipc_dai_dmic_params dmic;
	};
};
```

**Comment by ranj063** [comment]
@xiulipan has this dai_get() function changed recently? 

How did the DMIC's work before?

**Comment by ranj063** [comment]
We need @singalsu to comment too. I think you are probably mixnig up FIFO A and FIFOB with DMIC0 and DMIC1. Not sure if it is correct.

**Comment by xiulipan** [comment]
@ranj063 

> How did the DMIC's work before?

Luckily we only used DMIC in topology before, no SSP involved, so the id and index is both 0 in this situation.
But in your patch, the id is 6. But the DAI INDEX for DMIC is still 0 from the DMIC structure in src/platform/apollolake/dai.c

**Comment by ranj063** [comment]
@xiulipan could we please test a test topology with just the DMIC DAI?
Have you got it working on the up2 yet?

**Comment by ranj063** [comment]
@xiulipan @plbossart I've made the changes to address the 16k capture for DMIC1 and fix the id's for the DMIC's. Could you please review the new version?

File: topology/sof-apl-nocodec.m4 [hunk 1]
```diff
@@ -249,3 +280,13 @@ DAI_CONFIG(SSP, 5, 5, NoCodec-5,
 		      SSP_TDM(2, 16, 3, 3),
 		      SSP_CONFIG_DATA(SSP, 5, 16)))
 
+DAI_CONFIG(DMIC, 6, 6, NoCodec-6,
+	   DMIC_CONFIG(1, 500000, 4800000, 40, 60, 48000,
+		DMIC_WORD_LENGTH(s32le), DMIC, 6,
+		PDM_CONFIG(DMIC, 6, FOUR_CH_PDM0_PDM1)))
```
Comment (xiulipan) [inline topology/sof-apl-nocodec.m4]:
```
DAI_CONFIG(DMIC, 0, 6, NoCodec-6,
	   DMIC_CONFIG(1, 500000, 4800000, 40, 60, 48000,
		DMIC_WORD_LENGTH(s32le), DMIC, 0,
		PDM_CONFIG(DMIC, 0, FOUR_CH_PDM0_PDM1)))
```

File: topology/sof-apl-nocodec.m4 [hunk 2]
```diff
@@ -249,3 +280,13 @@ DAI_CONFIG(SSP, 5, 5, NoCodec-5,
...
+
+DAI_CONFIG(DMIC, 7, 7, NoCodec-7,
+	   DMIC_CONFIG(1, 500000, 4800000, 40, 60, 16000,
+		DMIC_WORD_LENGTH(s32le), DMIC, 7,
+		PDM_CONFIG(DMIC, 7, FOUR_CH_PDM0_PDM1)))
```
Comment (xiulipan) [inline topology/sof-apl-nocodec.m4]:
```
DAI_CONFIG(DMIC, 1, 7, NoCodec-7,
	   DMIC_CONFIG(1, 500000, 4800000, 40, 60, 16000,
		DMIC_WORD_LENGTH(s32le), DMIC, 1,
		PDM_CONFIG(DMIC, 1, FOUR_CH_PDM0_PDM1)))
```

**Comment by xiulipan** [comment]
Find error message for DMIC 1 config:

[    3.445830] sof-audio sof-audio: tplg: config DMIC1 driver version 1
[    3.445832] sof-audio sof-audio: pdmclk_min 500000 pdm_clkmax 4800000 duty_min 40
[    3.445834] sof-audio sof-audio: duty_max 60 fifo_fs 48000 num_pdms active 2
[    3.445835] sof-audio sof-audio: fifo word length 32
[    3.445837] sof-audio sof-audio: pdm 0 mic a 1 mic b 1
[    3.445839] sof-audio sof-audio: pdm 0 polarity a 0 polarity b 0
[    3.445840] sof-audio sof-audio: pdm 0 clk_edge 0 skew 0
[    3.445842] sof-audio sof-audio: pdm 1 mic a 1 mic b 1
[    3.445843] sof-audio sof-audio: pdm 1 polarity a 0 polarity b 0
[    3.445845] sof-audio sof-audio: pdm 1 clk_edge 0 skew 0
[    3.445865] sof-audio sof-audio: ipc: send 0x80010000
[    3.445961] sof-audio sof-audio: error: ipc error for 0x80010000 size 0xc
[    3.446053] sof-audio sof-audio: error: failed to set DAI config for DMIC1

rmbox log:
Error log:
using 19.20MHz timestamp clock
0x0 [0.095826]  delta [0.095826]        dmic cfr
0x10 [0.095828] delta [0.000001]        dmic cor
0x20 [0.095830] delta [0.000002]        ipc eDC

Seems we did not probe the DMIC?

**Comment by xiulipan** [comment]
Fixed dmic probe with https://github.com/thesofproject/sof/pull/141

**Comment by xiulipan** [comment]
DMIC1 could not work.
Seems we did no test with it
@ranj063 @lgirdwood 
do you have any idea about if dmic1 is enabled and tested? Should we enable DMIC1 here without test before?
I did not know what is happening here.

rmbox full of trace like:
0x1610 [116.457545]     delta [0.000004]        dmic eor
0x1620 [116.457550]     delta [0.000005]        dmic eor
0x1630 [116.457554]     delta [0.000004]        dmic eor
0x1640 [116.457559]     delta [0.000005]        dmic eor
0x1650 [116.457563]     delta [0.000004]        dmic eor
0x1660 [116.457568]     delta [0.000005]        dmic eor
0x1670 [116.457572]     delta [0.000004]        dmic eor
0x1680 [116.457577]     delta [0.000005]        dmic eor
0x1690 [116.457581]     delta [0.000004]        dmic eor
0x16a0 [116.457586]     delta [0.000005]        dmic eor
0x16b0 [116.457590]     delta [0.000004]        dmic eor
0x16c0 [116.457595]     delta [0.000005]        dmic eor
0x16d0 [116.457599]     delta [0.000004]        dmic eor
0x16e0 [116.457604]     delta [0.000005]        dmic eor
0x16f0 [116.457608]     delta [0.000004]        dmic eor
0x1700 [116.459892]     delta [0.002285]        dmic eor

**Comment by singalsu** [comment]
Why is the PCM7 DMIC called DMIC16k?

**Comment by ranj063** [comment]
@singalsu thanks for the clarification 

DMIC16k name is coming from the machine driver.

**Comment by ranj063** [comment]
@plbossart @xiulipan the dai definitions in the topology need a bit of cleaning up. 

Currently, we have id, link_id and dai_index. But I think we should really have port, link_id, and dai_index. link_id is what is matched against the machine driver and dai_index is what is  used in the FW. 

I;m not sure if port number is appropriate for the DMIC though.

**Comment by xiulipan** [comment]
@singalsu 
Good to know that is not some unknown bug but an unfinished job.
Then we are good to go with DMIC01 only.

File: topology/sof-apl-nocodec.m4 [hunk 3]
```diff
@@ -211,23 +221,28 @@ DAI_ADD(sof/pipe-dai-capture.m4,
 	48, 1000, 0, 0)
 
 # PCM Low Latency, id 0
+dnl PCM_DUPLEX_ADD(name, pipeline, pcm_id, dai_id, playback, capture)
```
Comment (plbossart) [inline topology/sof-apl-nocodec.m4]: dai_id: did you mean link_id?

File: topology/sof-apl-nocodec.m4 [hunk 4]
```diff
@@ -211,23 +221,28 @@ DAI_ADD(sof/pipe-dai-capture.m4,
...
 PCM_DUPLEX_ADD(Port4, 4, 4, 4, PIPELINE_PCM_9, PIPELINE_PCM_10)
 PCM_DUPLEX_ADD(Port5, 5, 5, 5, PIPELINE_PCM_11, PIPELINE_PCM_12)
+dnl PCM_CAPTURE_ADD(name, pipeline, pcm_id, dai_id, capture)
 PCM_CAPTURE_ADD(DMIC01, 6, 6, 6, PIPELINE_PCM_13)
 
 #
 # BE configurations - overrides config in ACPI if present
 #
 
+dnl DAI_CONFIG(type, index, link_id, name, ssp_config/dmic_config)
```
Comment (plbossart) [inline topology/sof-apl-nocodec.m4]: type, dai_index, link_id, link_name?

File: topology/sof-apl-nocodec.m4 (hunk 3 above)
Comment (ranj063) [reply topology/sof-apl-nocodec.m4]: @plbossart this is actually not the link_id. 
But I'm making some more clean up for this. 

I think we should avoid the dai_id in the pcm definition. When creating cpu_dai's for the FE dai's, the id should come from the component->num_dai just like the cpu_dai id for the BE DAI's.

My next pull request will be about this change.

**Comment by ranj063** [comment]
@plbossart I've updated the comments to reflect the changes and yes this patch depends on these:
https://github.com/thesofproject/soft/pull/40
https://github.com/thesofproject/linux/pull/42
https://github.com/thesofproject/sof/pull/140

**Comment by plbossart** [comment]
what about soft #42, isn't that a dependency as well?

**Comment by ranj063** [comment]
@plbossart #42 isnt a dependency for this to work. 

#42 just makes sure we use the right ID instead of index which is the right thing to do IMO.

But this topology would work even without that change.

**Comment by xiulipan** [comment]
LGTM, thanks

**Comment by lgirdwood** [comment]
@ranj063 cant find the patches it appears to be missing.

**Comment by plbossart** [comment]
@ranj063 I don't get the logic here: 
"#42 just makes sure we use the right ID instead of index which is the right thing to do IMO.
But this topology would work even without that change."

How can the topology work with the wrong ID?

**Comment by ranj063** [comment]
@plbossart sorry i was confused. You are right #42 will be needed for DMIC to work.

**Comment by ranj063** [comment]
@lgirdwood I want to update the PIPELINE_PCM_DAI_ADD with PIPELINE_PCM_ADD in the topology too.
I will send you a notification once I make that change.

**Comment by ranj063** [comment]
@lgirdwood @plbossart I've updated this PR to replace pipeline_pcm_dai_add with pipeline_pcm_add

**Comment by ranj063** [comment]
@plbossart I think this pull request is awaiting your approval. Could you please ack?

**Pull request merged**